File: src/main/java/org/entities/aggressive/TallGolem.java
package org.entities.aggressive;

import org.engine.Main;
import org.entities.Autonomous;
import org.entities.SmartRectangle;
import org.graphics.Animator;
import org.graphics.Graphics;
import org.level.Level;
import org.level.LevelController;
import org.loader.ImageResource;
import org.loader.ResourceHandler;
import org.world.*;

public class TallGolem extends Autonomous {
    public static final int BLUE=0;
    private int golemType;
    private Animator golemAnimator;
    private SmartRectangle hitbox=new SmartRectangle(x,y,width,height);
    private ImageResource golem;
    public TallGolem(int golemType, int subLevel, float spawnX, float spawnY) {
        super(subLevel, spawnX, spawnY);
        this.golemType=golemType;
        switch(golemType){
            case BLUE:
                health=3;
                displayName="Tall Blue Golem";
                golemAnimator=new Animator(ResourceHandler.getGolemLoader().getTallBlueGolemWalk(direction),10);
                break;
        }
    }

    @Override
    public void update() {
        //Get HeightMap information
        HeightReturn h= HeightMap.onGround(hitbox);

        //Movement input
        if(damageTakenFrame==0) {
            if (direction) {
                vX = .3f;
            } else {
                vX = -.3f;
            }
        }else{
            if((direction&&!attackerBehind)||(!direction&&attackerBehind))vX=-.8f;
            else vX=.8f;
            damageTakenFrame--;
        }

        if(damageCooldown>0){
            damageCooldown--;
            state=2;
        }else if(state==2){
            state=0;
        }

        //Y-Velocity Calculations
        y+=vY;
        vY-= World.getGravity();

        //X-Velocity and Jumping Calculations
        boolean doXCalc=true;

        if (HeightMap.checkRightCollision(hitbox)) {
            HeightVal hv=HeightMap.findApplicable(hitbox,true);
            if (hv!=null&&x + width + 0.5>= hv.getStartX()) {
                if (vX < 0) x += vX;
                else vX=0;
                doXCalc=false;
                direction=!direction;
                x-=.25f;
            }
        }
        if(HeightMap.checkLeftCollision(hitbox)){
            HeightVal hv=HeightMap.findApplicable(hitbox,false);
            if(hv!=null&&x-0.5<=hv.getEndX()){
                if(vX>0)x+=vX;
                else vX=0;
                doXCalc=false;
                direction=!direction;
                x+=.25f;
            }
        }
        HeightVal wall=HeightMap.findNextWall(hitbox,direction);
        final int jumpDist=5;
        if(wall !=null&&y+32>=wall.getHeight()) {
            if (direction) {
                if (wall.getStartX() - x - width <= jumpDist) {
                    state = 3;
                }
            } else {
                if (x - wall.getEndX() <= jumpDist) {
                    state = 3;
                }
            }
        }
        if(state==3&&golemAnimator.getCurrentFrameNum()<golemAnimator.getFrames().length-1)doXCalc=false;
        if(doXCalc){
            x+=vX;
            doXCalc();
        }

        //Y-velocity and ground calc
        if(h.isOnGround()&&vY<0){
            y=h.getGroundLevel();
            vY=0;
            if(state==3&&golemAnimator.getCurrentFrameNum()==golemAnimator.getFrames().length-1)
                state=0;
        }

        doAttackCalc();

        //Update the frames
        if(vX==0){
            switch (golemType){
                case BLUE:
                    golemAnimator.setFrames(new ImageResource[]{ResourceHandler.getGolemLoader().getTallBlueGolem(direction)});
                    break;
            }
        }else {
            doSpriteCalc();
        }

        if (state != 3 || golemAnimator.getCurrentFrameNum() < golemAnimator.getFrames().length - 1) {
            golemAnimator.update();
        }

        if(state==1&&golemAnimator.getCurrentFrameNum()==golemAnimator.getFrames().length-1){
            state=0;
            Attack.attack(this,1,4);
        }
        if(state==3&&golemAnimator.getCurrentFrameNum()==golemAnimator.getFrames().length-1){
            if(vY==0){
                vY=3f;
            }
        }
    }

    private void doSpriteCalc(){
        golem = golemAnimator.getCurrentFrame();
        switch (state) {
            case 0:
                switch (golemType) {
                    case BLUE:
                        golemAnimator.setFrames(ResourceHandler.getGolemLoader().getTallBlueGolemWalk(direction));
                        break;
                }
                break;
            case 1:
                switch (golemType) {
                    case BLUE:
                        golemAnimator.setFrames(ResourceHandler.getGolemLoader().getTallBlueGolemAttack(direction));
                        break;
                }
                break;
            case 2:
                switch (golemType) {
                    case BLUE:
                        golemAnimator.setFrames(new ImageResource[]{ResourceHandler.getGolemLoader().getTallBlueGolemKnockback(direction)});
                        break;
                }
                break;
            case 3:
                switch (golemType) {
                    case BLUE:
                        golemAnimator.setFrames(ResourceHandler.getGolemLoader().getTallBlueGolemJump(direction));
                        break;
                }
                break;
        }
    }

    private void doXCalc(){
        Level l= LevelController.getCurrentLevel();
        if(x<l.getLeftLimit()||x+width>l.getRightLimit()){
            direction=!direction;
        }
    }

    private void doAttackCalc(){
        if(state!=0)return;
        if(attackCooldown>0){
            attackCooldown--;
            return;
        }
        if(Main.getHarold().getY()>y+height||Main.getHarold().getY()+Main.getHarold().getWidth()<y){
            state=0;
            return;
        }
        if(direction){
            if(Main.getHarold().getX()>=x&&Main.getHarold().getX()<=x+width+6){
                state=1;
                golemAnimator.setCurrentFrame(0);
                attackCooldown=100;
            }
        }else{
            if(Main.getHarold().getX()+Main.getHarold().getWidth()<=x+width&&Main.getHarold().getWidth()+Main.getHarold().getX()>=x-6){
                state=1;
                golemAnimator.setCurrentFrame(0);
                attackCooldown=100;
            }
        }
    }

    @Override
    public void render() {
        if(golem==null)return;
        //Update hitbox info
        width= Graphics.convertToWorldWidth(golem.getTexture().getWidth());
        height=Graphics.convertToWorldHeight(golem.getTexture().getHeight());
        hitbox.updateBounds(x,y,width,height);

        //Draw the golem
        if(damageTakenFrame>0){
            Graphics.setColor(1,0,0,1);//Set damage color if needed
        }
        else Graphics.setColor(1,1,1,1);
        Graphics.drawImage(golem,x,y);
        Graphics.setColor(1,1,1,1);
    }

    @Override
    public void reset() {

    }

    @Override
    public String toString() {
        switch(golemType) {
            case 0:
                return "Tall Blue Golem @ " + x + "," + y;
            default:
                return "Tall Golem @ "+x+","+y;
        }
    }
}

File: src/main/java/org/entities/aggressive/TallGolem.kt
package org.entities.aggressive

import org.engine.Main
import org.entities.Autonomous
import org.entities.SmartRectangle
import org.graphics.Animator
import org.graphics.Graphics
import org.level.LevelController.currentLevel
import org.loader.ImageResource
import org.loader.ResourceHandler
import org.world.Attack.melee
import org.world.HeightMap
import org.world.World.gravity

class TallGolem(private val golemType: Int, subLevel: Int, spawnX: Float, spawnY: Float) :
    Autonomous(subLevel, spawnX, spawnY) {
    private var golemAnimator: Animator? = null
    private val hitbox = SmartRectangle(x, y, width, height)
    private var golem: ImageResource? = null
    override fun update(deltaTime: Float) {
        //Get HeightMap information
        val h = HeightMap.onGround(hitbox)

        //Movement input
        if (damageTakenFrame == 0f) {
            vX = if (direction) {
                18f
            } else {
                -18f
            }
        } else {
            vX = if (direction && !attackerBehind || !direction && attackerBehind) -.8f else 12f
            damageTakenFrame--
        }
        if (damageCooldown > 0) {
            damageCooldown--
            state = 2
        } else if (state == 2) {
            state = 0
        }

        //Y-Velocity Calculations
        y += vY * deltaTime
        vY -= gravity * deltaTime * 2

        //X-Velocity and Jumping Calculations
        var doXCalc = true
        if (HeightMap.checkRightCollision(hitbox)) {
            val hv = HeightMap.findApplicable(hitbox, true)
            if (hv != null && x + width + 0.5 >= hv.startX) {
                if (vX < 0) x += vX * deltaTime else vX = 0f
                doXCalc = false
                direction = !direction
                x -= .25f
            }
        }
        if (HeightMap.checkLeftCollision(hitbox)) {
            val hv = HeightMap.findApplicable(hitbox, false)
            if (hv != null && x - 0.5 <= hv.endX) {
                if (vX > 0) x += vX * deltaTime else vX = 0f
                doXCalc = false
                direction = !direction
                x += .25f
            }
        }
        val wall = HeightMap.findNextWall(hitbox, direction)
        val jumpDist = 7
        if (wall != null && y + 32 >= wall.height) {
            if (direction) {
                if (wall.startX - x - width <= jumpDist) {
                    state = 3
                }
            } else {
                if (x - wall.endX <= jumpDist) {
                    state = 3
                }
            }
        }
        if (state == 3 && golemAnimator!!.currentFrameNum < golemAnimator!!.frames.size - 1) doXCalc = false
        if (doXCalc) {
            x += vX * deltaTime
            doXCalc()
        }

        //Y-velocity and ground calc
        if (h.isOnGround && vY < 0) {
            y = h.groundLevel
            vY = 0f
            if (state == 3 && golemAnimator!!.onLastFrame()) state = 0
        }
        doAttackCalc()

        //Update the frames
        if (vX == 0f) {
            when (golemType) {
                BLUE -> golemAnimator!!.frames = arrayOf(ResourceHandler.getGolemLoader().getTallBlueGolem(direction))
            }
        } else {
            doSpriteCalc()
        }
        if (state != 3 || golemAnimator!!.currentFrameNum < golemAnimator!!.frames.size - 1) {
            golemAnimator!!.update()
        }
        if (state == 1 && golemAnimator!!.onLastFrame()) {
            state = 0
            melee(this, 1, 4f)
        }
        if (state == 3 && golemAnimator!!.onLastFrame()) {
            if (vY == 0f) {
                vY = 110f
            }
        }
    }

    private fun doSpriteCalc() {
        golem = golemAnimator!!.currentFrame
        when (state) {
            0 -> when (golemType) {
                BLUE -> golemAnimator!!.frames = ResourceHandler.getGolemLoader().getTallBlueGolemWalk(direction)
            }
            1 -> when (golemType) {
                BLUE -> golemAnimator!!.frames = ResourceHandler.getGolemLoader().getTallBlueGolemAttack(direction)
            }
            2 -> when (golemType) {
                BLUE -> golemAnimator!!.frames =
                    arrayOf(ResourceHandler.getGolemLoader().getTallBlueGolemKnockback(direction))
            }
            3 -> when (golemType) {
                BLUE -> golemAnimator!!.frames = ResourceHandler.getGolemLoader().getTallBlueGolemJump(direction)
            }
        }
    }

    private fun doXCalc() {
        val l = currentLevel
        if (x < l.leftLimit || x + width > l.rightLimit) {
            direction = !direction
        }
    }

    private fun doAttackCalc() {
        if (state != 0) return
        if (attackCooldown > 0) {
            attackCooldown--
            return
        }
        if (Main.getHarold().y > y + height || Main.getHarold().y + Main.getHarold().width < y) {
            state = 0
            return
        }
        if (direction) {
            if (Main.getHarold().x >= x && Main.getHarold().x <= x + width + 6) {
                state = 1
                golemAnimator!!.setCurrentFrame(0)
                attackCooldown = 100f
            }
        } else {
            if (Main.getHarold().x + Main.getHarold().width <= x + width && Main.getHarold().width + Main.getHarold().x >= x - 6) {
                state = 1
                golemAnimator!!.setCurrentFrame(0)
                attackCooldown = 100f
            }
        }
    }

    override fun render() {
        if (golem == null) return
        //Update hitbox info
        width = Graphics.toWorldWidth(golem!!.texture.width.toFloat())
        height = Graphics.toWorldHeight(golem!!.texture.height.toFloat())
        hitbox.updateBounds(x, y, width, height)

        //Draw the golem
        if (damageTakenFrame > 0) {
            Graphics.setDrawColor(1f, 0f, 0f, 1f) //Set damage color if needed
        } else {
            Graphics.setDrawColor(1f, 1f, 1f, 1f)
        }
        Graphics.drawImage(golem!!, x, y)
        Graphics.setDrawColor(1f, 1f, 1f, 1f)
    }

    override fun reset() {}
    override fun toString(): String {
        return when (golemType) {
            0 -> "Tall Blue Golem @ $x,$y"
            else -> super.toString()
        }
    }

    companion object {
        const val BLUE = 0
    }

    init {
        when (golemType) {
            BLUE -> {
                health = 3
                displayName = "Tall Blue Golem"
                golemAnimator = Animator(ResourceHandler.getGolemLoader().getTallBlueGolemWalk(direction), 10)
            }
        }
    }
}